tiledb/ml/data_apis/pytorch_sparse.py

```diff
@@ -0,0 +1,122 @@
+        :param batch_size: Integer. The size of the batch that the generator will return. Remember to set batch_size=None
+        when calling the PyTorch Sparse Dataloader API, because batching is taking place inside the TileDB IterableDataset.
+        """
+
+        # Check that x and y have the same number of rows
+        if x_array.schema.domain.shape[0] != y_array.schema.domain.shape[0]:
+            raise ValueError(
+                "X and Y should have the same number of rows, i.e., the 1st dimension "
+                "of TileDB arrays X, Y should be of equal domain extent."
+            )
+
+        self.x = x_array
+        self.y = y_array
+        self.batch_size = batch_size
+
+        if isinstance(y_array, tiledb.SparseArray):
+            self.iter_sparse_label = True
+        else:
+            self.iter_sparse_label = False
+
+    def __iter__(self):
+        worker_info = torch.utils.data.get_worker_info()
+
+        # Get number of observations
+        rows = self.x.schema.domain.shape[0]
+
+        # Single worker - return full iterator
+        if worker_info is None:
+            iter_start = 0
+            iter_end = rows
+
+        # Multiple workers - split workload
+        else:
+            per_worker = int(math.ceil(rows / worker_info.num_workers))
+            worker_id = worker_info.id
+            iter_start = worker_id * per_worker
+            iter_end = min(iter_start + per_worker, rows)
+
+        y_attr_name = self.y.schema.attr(0).name
+
+        x_shape = self.x.schema.domain.shape[1:]
+        y_shape = self.y.schema.domain.shape[1:]
+
+        # Loop over batches
+        for offset in range(iter_start, iter_end, self.batch_size):
+            # Yield the next training batch
+            y_batch = self.y[offset : offset + self.batch_size]
+            x_batch = self.x[offset : offset + self.batch_size]
+
+            # TODO: Both for dense case support multiple attributes
+            values_y = list(y_batch.items())[0][1]
+            values_x = list(x_batch.items())[0][1]
+
+            # Transform to TF COO format y data
+            y_data = np.array(values_y).ravel()
+            x_data = np.array(values_x).ravel()
+            if x_data.shape[0] != y_data.shape[0]:
+                raise ValueError(
+                    "X and Y should have the same number of rows, i.e., the 1st dimension "
+                    "of TileDB arrays X, Y should be of equal domain extent inside the batch"
+                )
+
+            x_coords = []
+            for i in range(0, self.x.schema.domain.ndim):
+                dim_name = self.x.schema.domain.dim(i).name
+                x_coords.append(np.array(x_batch[dim_name]))
+
+            # Normalise indices for torch.sparse.Tensor are batched with 0-index
+            x_coords[0] = np.vectorize(
+                lambda x: x - np.max(x_coords[0]) + self.batch_size - 1
+            )(x_coords[0])
```

Comment:
This is equivalent to `x_coords[0] -= x_coords[0].max() - self.batch_size + 1`, no need for `np.vectorize`.

Also are you sure about using `max()` here instead of `min()`? Seems like there can be negative coords with `max()`.

Reply:
I have tested the functionality for supporting positive integer `idx`. In fact using this is allowed due to the fact that `x_coords` are result of slicing` x_array [offset:offset+batch_size]` with offsets being in the range of `[0, ...)`. For now we support the basic scenarios of domains, mostly used. However I will put a comment for further investigation in case we want to support either string, float or negative domains.